python/mim_estimation/ekf.py

```diff
@@ -1,243 +1,262 @@
+
     # progagate covariance 
-    def prediction_step(self, Fk, Qk):
-        self.__Sigma = np.dot(np.dot(Fk, self.__Sigma), Fk.T) + Qk 
-    
-    def predict_measurements(self):
-        z = np.zeros(12, dtype=float)
-        C = self.get_base_orientation().rotation_matrix
-        z[0:3] = np.dot(C, (self.get_p1_position()-self.get_base_position()))
-        z[3:6] = np.dot(C, (self.get_p2_position()-self.get_base_position()))
-        z[6:9] = np.dot(C, (self.get_p3_position()-self.get_base_position()))
-        z[9:12] = np.dot(C, (self.get_p4_position()-self.get_base_position()))
-        return z
+    def prediction_step(self):
+        Fk = self.compute_discrete_prediction_jacobian()
+        Lc = self.compute_noise_jacobian()
+        Qc = self.construct_continuous_noise_covariance()
+        Qk = self.construct_discrete_noise_covariance(Fk, Lc, Qc)
+        self.__Sigma_pre = (Fk @ self.__Sigma_post @ Fk.T) + Qk 
     
-    def get_new_measurments(self, robot_configuration):
-        s = np.zeros(12, dtype=float)
-        end_effectors_positions = self.__compute_end_effectors_positions(robot_configuration)
-        C = self.__compute_base_pose_se3(robot_configuration).rotation_matrix
-        r = C.translation 
-        s[0:3] = np.dot(C, (end_effectors_positions['p1']-r))
-        s[3:6] = np.dot(C, (end_effectors_positions['p2']-r))
-        s[6:9] = np.dot(C, (end_effectors_positions['p3']-r))
-        s[9:12] = np.dot(C, (end_effectors_positions['p4']-r))
-        return s
+    # assuming contact logic variables coming from the contact schedule for now
+    # TODO contact logic prediction using a probabilistic model
+    def measurement_model(self, contacts_schedule, joint_positions, joint_velocities):
+        Hk = np.zeros((12, self.__nx))  #12x15
+        predicted_base_velocity = np.zeros(12)
+        measured_base_velocity = np.zeros(12)
+        dt = self.__dt
+        r_pre = self.__mu_pre['base_position']
+        q_pre = self.__mu_pre['base_orientation']
+        R_pre = pin.XYZQUATToSE3([r_pre[0], r_pre[1], r_pre[2],
+                q_pre.x, q_pre.y, q_pre.z, q_pre.w]).rotation
+        r_post = self.__mu_post['base_position']
+        q_post = self.__mu_post['base_orientation']
+        R_post = pin.XYZQUATToSE3([r_post[0], r_post[1], r_post[2],
+                 q_post.x, q_post.y, q_post.z, q_post.w]).rotation
+        # end effectors frame positions and velocities expressed in the world frame 
+        ee_placements, ee_velocities = self.compute_end_effectors_FK_quantities(joint_positions, joint_velocities)
+        # compute measurement jacobian
+        Hk[0:3,3:6] = Hk[3:6,3:6] = Hk[6:9,3:6] = Hk[9:12,3:6] = -np.eye(3)
+        Hk[0:3,9:12] = -R_pre @ pin.skew(ee_placements['FL'])
+        Hk[3:6,9:12] = -R_pre @ pin.skew(ee_placements['FR'])
+        Hk[6:9,9:12] = -R_pre @ pin.skew(ee_placements['HL'])
+        Hk[9:12,9:12] = -R_pre @ pin.skew(ee_placements['HR'])
```

Comment:
Looking back at equation (15) in [your derivations](https://github.com/machines-in-motion/mim_estimation/issues/7#issuecomment-760354409), I don't see where the R_pre terms multiplying the skew symmetric foot positions are coming from. Did you find an error in this derivation?

Reply:
Thanks for catching this. Looking back at equation (11), the foot placements are expressed in the base frame. This means the rotation matrix has to be perturbed as well and have an effect in the analytic jacobian. This was not visible when doing finite differencing as I also didn't consider this rotation matrix when computing $f(x+\delta)$. May be this is the missing piece in the measurement jacobian. Now, I am also doubting perturbing the foot velocity term of the feet. I will redo this computation and post it back.

Reply:
So I revisited this again, and I guess I get the source of bug. From the above snippet, I wanted to bring the foot placements in the base frame, which means R_pre needs to be inside the cross product such that it's coherent with equation (15) as all quantities are computed in the base frame there. I am pushing a commit now to fix this bug. It doesn't mean though that I need to perturb this rotation matrix (i.e. what I said in the previous comment is wrong). @nrotella Does that make sense?

Reply:
I think that does make sense - the foot position (placement) relative to the base in base frame only depends on kinematics/joint angles, so you shouldn't need to perturb the rotation matrix, I think.

There is still something a little confusing about the derivation of the measurement Jacobian, though. Right now, the only nonzero terms are on the gyro bias, which means the measurement can only affect that part of the state directly. I would think there'd be some information from the measurement which affects the velocity and/or orientation, but I can't figure out where those would come in.

I will think about it some more later today, but for now, can you check the measurement Jacobian numerically now that the bug is fixed? Does it agree with the analytic Jacobian?

Reply:
There is also another term depending on the velocity as well considering my old derivation. I think perturbing the velocity of the foot makes sense since it is a function of the velocity of the base. The way I think about is picturing a squatting motion, the velocity of the foot w.r.t. the base in the base frame is changing based on the velocity of the base, unless I am confusing something of course.

Reply:
I recall you expanding the foot velocity as if it were the base velocity in the old derivation, but that didn't quite seem correct either. The velocity of the foot relative to the base _in the base frame_ should be a function of only the kinematic model and joint state, no? I'll try to think about this a little more over the weekend.

Did you get the chance to check the Jacobian numerically, though? As long as you have the measurement equation implemented correctly using the state, then any problems should reveal themselves as a mismatch in the Jacobians.

Reply:
@nrotella Yes, I checked against the numerical measurement jacobian. However, in this check I perturbed the velocity of the base inside the velocity of the foot term as well during finite differencing, which is not a good check in this case for catching this bug, as in  both the analytical and numerical jacobians I perturbed the same term (which seem wrong even when they match in this case). For the portion of the angular velocity bias it's correct for sure. In the end, if I remove this velocity term, I don't see any contribution in the measurement jacobian except from the angular velocity bias term.

Reply:
The way I understand the relative velocity equations is like this:
Velocity of a point $A$ is equal to the velocity of $B$ plus the relative velocity and when we have a coordinate frame that is attached to B, the relative velocity is a summation of the velocity due to rotation of the frame plus the velocity in the frame. It reads:
\begin{equation}
v_A = v_B + \omega \times r_{BA} + \dot r_{BA} 
\end{equation}
This equation is invariant to the coordinate frame, i.e. it is valid either we express all the vectors in the local frame or global frame (I checked this many times:) ). In our case, we can write the velocity equation between base and foot either in body or world frame. Since we are interested in writing all the values in the body frame, we have:
\begin{equation}
0 = v_K = v_B + \omega \times r_{BK} + \dot r_{BK} 
\end{equation}
Note that all the values are expressed in the body frame, so I dropped frames for all the values. In this equation $K$ stands for the foot in contact. In the right-hand-side of this equation, the first term is the velocity of the base expressed in the base frame, the second term is the cross product of the base angular velocity expressed in the body frame and the position of the foot with respect to the base in the base frame (this can be obtained by multiplying of the transformation matrix starting from the base to the foot). Finally the last term is the velocity of the foot with respect to the base in the base frame which can be obtained using Jacobian of the foot (the first six columns are excluded) multiplied to the joint velocities. This is our measurement model that we can use to find the Jacobian.

Reply:
This all makes sense to me now, thanks for the detailed comment! I was getting confused by the way the measurement is formulated in Pronto:

\begin{equation}
v_{WB}^{B} = -v_{BK}^{B} - \omega_{WB}^{B} \times r_{BK}^{B}
\end{equation}

which implies that the base velocity in base frame $v_{WB}^{B}$ is what we're measuring, but what we should be measuring is the foot velocity relative to world in base frame (which is zero) as you state above. moving this base velocity term to the right and negating, we have:
 
\begin{equation}
0 = v_{WK}^{B} = v_{WB}^{B}  + v_{BK}^{B} + \omega_{WB}^{B} \times r_{BK}^{B}
\end{equation}

This is also consistent with [Bloesch et al](https://www.research-collection.ethz.ch/bitstream/handle/20.500.11850/75852/eth-7743-01.pdf) eq (31), taking into account that their velocity state is actually the negative of base velocity in base frame (for whatever reason). I think the way the [Pronto paper](https://github.com/machines-in-motion/mim_estimation/files/5789205/pronto.pdf) states the measurement was the source of confusion, for me at least. The Bloesch equation makes sense.

So, we have:

Reply:
Yes, this makes perfect sense to me as well now. Thanks  @majidkhadiv for clarifying this point !